src/functions.php

```diff
@@ -15,13 +15,13 @@
  *
  * @param callable(mixed ...$args):mixed $function
  * @param mixed ...$args Optional list of additional arguments that will be passed to the given `$function` as is
- * @return PromiseInterface<mixed>
+ * @return callable(): PromiseInterface<mixed>
```

Comment:
The `$args` parameter should be omitted here and below.

How about this?

```
 * @param callable(mixed ...$args): mixed $function
 * @return callable(mixed ...$args): PromiseInterface<mixed>
```

Reply:
This is actually unsafe from a static analysis perspective, it would be better to get rid of it completely.

Reply:
@azjezz Why is it unsafe from a static analysis perspective?

Reply:
$callable = function(int $a): void {}
$async_callable = async($callable, new User()); // <-- both psalm, and phpstan allow this, but it's wrong.

Reply:
and passing outside arguments is not hard in PHP 8.1, `async(fn() => foo($a, $b, $c))`

Reply:
@azjezz I'm not sure I follow. With these changes, the `async()` function no longer accepts any arguments, but it returns a new function that accepts any number of arguments that will be forwarded as-is to the original `$function`. Are you aware of a better type definition for this?

Reply:
no, there's no way to actually type this in PHP ( and most other typed languages ).

async is currently: (F(...unknown) -> Y) + ...unknown -> (F() -> Promise<Y>)
your suggestion: (F(...unknown) -> Y) -> (F(...unknown) -> Promise<Y>)  

The issue is not just that we don't know type of 'unknown', but also how many.

so even this:

```php
$async = async(fn(string $identifier): User => $db->getUser($identifier));

$result = await($async([1, 2, 3]));
```

will pass static analysis since the returned callable is annotated as `callable(mixed ...$args): Promise<mixed>`, but fails at runtime.

( aside from dropping $args, it would be nice to add generics so that the return type of callable is known, as of now, `$result` in the example above is detected as `mixed` while it should be `User` ) 

ref: https://github.com/azjezz/psl/blob/2.0.x/src/Psl/Async/run.php#L15-L19

Reply:
@azjezz I think we're on the same page here. I think we agree that the `@param mixed ...$params` should be removed and `@return callable(): PromiseInterface<mixed>` is probably good enough as a return type until we see better support for this in static analysis tools :+1: